river/cluster/textclust.py

```diff
@@ -0,0 +1,643 @@
+            if (
+                self.micro_clusters[key].weight <= self.omega
+                or len(self.micro_clusters[key].tf) == 0
+            ):
+                del self.micro_clusters[key]
+
+    # cleanup procedure
+    def _cleanup(self):
+
+        # set last cleanup to now
+        self.last_cleanup = self.t
+
+        # update current cluster weights
+        self._updateweights()
+
+        # set deltaweights
+        for micro in self.micro_clusters.values():
+
+            # here we compute delta weights
+            micro.deltaweight = micro.weight - micro.oldweight
+            micro.oldweight = micro.weight
+
+        # if auto merge is enabled, close micro clusters are merged together
+        if self.auto_merge:
+            self._mergemicroclusters()
+
+        ## reset merged observation
+        self._dist_mean = 0
+        self._num_merged_obs = 0
+
+    # merge
+    def _mergemicroclusters(self):
+        micro_keys = [*self.micro_clusters]
+
+        idf = self._calculateIDF(self.micro_clusters.values())
+        i = 0
+        if self.auto_r:
+            threshold = self._dist_mean / (self._num_merged_obs + 1)
+        else:
+            threshold = self.radius
+
+        while i < len(self.micro_clusters):
+            j = i + 1
+            while j < len(self.micro_clusters):
+                m_dist = self.micro_distance.dist(
+                    self.micro_clusters[micro_keys[i]], self.micro_clusters[micro_keys[j]], idf
+                )
+
+                ## lets merge them
+                if m_dist < threshold:
+                    self.micro_clusters[micro_keys[i]].merge(
+                        self.micro_clusters[micro_keys[j]],
+                        self.t,
+                        self.omega,
+                        self.fading_factor,
+                        self.term_fading,
+                        self.realtime,
+                    )
+                    del self.micro_clusters[micro_keys[j]]
+                    del micro_keys[j]
+                else:
+                    j = j + 1
+            i = i + 1
+
+    # calculate a distance matrix from all provided micro clusters
+    def _get_distance_matrix(self, clusters):
+
+        # if we need IDF for our distance calculation, we calculate it from the micro clusters
+        idf = self._calculateIDF(clusters.values())
+
+        # get number of clusters
+        numClusters = len(clusters)
+        ids = list(clusters.keys())
+
+        # initialize all distances to 0
+        distances = pd.DataFrame(np.zeros((numClusters, numClusters)), columns=ids, index=ids)
+
+        for idx, row in enumerate(ids):
+            for col in ids[idx + 1 :]:
+                # use the macro-distance metric to calculate the distances to different micro-clusters
+                dist = self.macro_distance.dist(clusters[row], clusters[col], idf)
+                distances.loc[row, col] = dist
+                distances.loc[col, row] = dist
```

Comment:
It's not efficient to call `.loc` many times like this. You should instead use simply Python lists, or a numpy array. And then convert the data to a pandas DataFrame. But I suppose we don't have to look into performance in this PR :)

Reply:
I agree. Not super efficient, but we should keep in mind that this method is only called when macro-clusters are generated from micro-clusters (which usually does not happen during the online phase). Therefore it is (a) really infrequently executed and (b) only used on the significantly smaller number of micro clusters.

Reply:
Understood. We could run some profiling to understand what could be improved anyway. But it's not important right now.